pkg/controller/clone-controller.go

```diff
@@ -102,162 +185,172 @@ func (c *CloneController) Run(threadiness int, stopCh <-chan struct{}) error {
-	if err != nil {
-		return errors.WithMessage(err, "could not retrieve pvc from cache")
+	// filter pvc and decide if the cloning pods should be created
+	if !checkClonePVC(pvc) {
+		return nil
 	}
-	// see if pvc's pod phase anno needs to be added/updated. The update is done only on the target PVC
-	phase := string(pod.Status.Phase)
-	_, exists := pvc.ObjectMeta.Annotations[AnnCloneRequest]
-	if !checkIfAnnoExists(pvc, AnnPodPhase, phase) && exists {
-		pvc, err = setPVCAnnotation(c.clientset, pvc, AnnPodPhase, phase)
-		if err != nil {
-			return errors.WithMessage(err, fmt.Sprintf("could not set annotation \"%s: %s\" on pvc %q", AnnPodPhase, phase, pvc.Name))
-
-		}
-		glog.V(Vdebug).Infof("processPodItem: pod phase %q annotated in pvc %q", pod.Status.Phase, pvcKey)
-		if phase == "Succeeded" {
-			pvc, err = setPVCAnnotation(c.clientset, pvc, AnnCloneOf, "true")
-			if err != nil {
-				return errors.WithMessage(err, fmt.Sprintf("could not set annotation \"%s: %s\" on pvc %q", AnnCloneOf, "true", pvc.Name))
-			}
-			glog.V(Vdebug).Infof("processPodItem: CloneOf annotatated in pvc %q", pvcKey)
-		}
-	}
-	return nil
+	glog.V(Vdebug).Infof("ProcessNextPvcItem: next pvc to process: %s\n", key)
+	return c.processPvcItem(pvc)
 }
 
 // Select only pvcs with the 'CloneRequest' annotation and that are not being processed.
 // We forget the key unless `processPvcItem` returns an error in which case the key can be
 // retried.
 func (c *CloneController) ProcessNextPvcItem() bool {
-	key, shutdown := c.pvcQueue.Get()
+	key, shutdown := c.queue.Get()
 	if shutdown {
 		return false
 	}
-	defer c.pvcQueue.Done(key)
+	defer c.queue.Done(key)
 
-	pvc, err := c.pvcFromKey(key)
-	if err != nil || pvc == nil {
-		return c.forgetKey(key, fmt.Sprintf("ProcessNextPvcItem: error converting key %v to pvc: %v", key, err))
-	}
-	// filter pvc and decide if the source and target pods should be created
-	if continue_processing, _, _ := checkClonePVC(c.clientset, pvc, false); !continue_processing {
-		return c.forgetKey(key, fmt.Sprintf("ProcessNextPvcItem: skipping pvc %q\n", key))
-	}
-	glog.V(Vdebug).Infof("ProcessNextPvcItem: next pvc to process: %s\n", key)
-	err = c.processPvcItem(pvc)
+	err := c.syncPvc(key.(string))
 	if err != nil { // processPvcItem errors may not have been logged so log here
 		glog.Errorf("error processing pvc %q: %v", key, err)
 		return true
 	}
 	return c.forgetKey(key, fmt.Sprintf("ProcessNextPvcItem: processing pvc %q completed", key))
 }
 
+func (c *CloneController) findClonePodsFromCache(pvc *v1.PersistentVolumeClaim) (*v1.Pod, *v1.Pod, error) {
+	selector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{MatchLabels: map[string]string{LabelClonePvc: pvc.Name}})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var sourcePod, targetPod *v1.Pod
+	annCloneRequest := pvc.GetAnnotations()[AnnCloneRequest]
+	if annCloneRequest != "" {
+		strArr := strings.Split(pvc.GetAnnotations()[AnnCloneRequest], "/")
```

Comment:
a helper function might be nice for the string parsing here. I see this code a few places.